File: Python/4sum.py
class Solution:
    def fourSum(self, arr: List[int], target: int) -> List[List[int]]:
        arr.sort()
        result=[]
        for j in range(len(arr)-3):
            if j > 0 and arr[j] == arr[j-1]:
                continue
            for i in range(j+1,len(arr)-2):
                if i > j+1 and arr[i] == arr[i-1]:
                    continue
                left=i+1
                right=len(arr)-1
                while left<right:
                    summ=arr[left]+arr[right]+arr[i]+arr[j]
                    if summ>target:
                        right-=1
                    elif summ==target:
                        result.append([arr[j], arr[i], arr[left], arr[right]])
                        while left<right and arr[left]==arr[left+1]:
                            left+=1
                        while left<right and arr[right]==arr[right-1]:
                            right-=1

                        left+=1
                        right-=1
                    else:
                        left+=1
        return result

File: Java/4Sum.java
 import java.util.*;
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        long total=0;
        Arrays.sort(nums);
        List<List<Integer>> arr = new ArrayList<>();
        for(int i =0 ; i < nums.length-3; i++){
             if (i>0 && nums[i]==nums[i-1]){
                    continue;
                }
            for ( int j =i+1 ; j < nums.length-2; j++){
                if (j>i+1 && nums[j]==nums[j-1]){
                    continue;
                }
                int left =j+1;
                int right=nums.length-1;
                while (left < right){
                    total=(long)nums[left]+nums[right]+nums[i]+nums[j];
                    if (total == target){
                        arr.add(Arrays.asList(nums[left],nums[right],nums[i],nums[j]));
                        while(left<right && nums[left]==nums[left+1]){
                        left++;
                        }
                        while(left<right && nums[right]==nums[right-1]){
                        right--;
                        }
                        left++;
                        right--;
                    }
                    else if(total>target){
                        right--;
                    }
                    else{
                        left++;
                    }
                    
                }

            }

        }
        return arr;
        
    }
}
